omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_9/ConceptReferenceTermSearchHandler1_9.java

```diff
@@ -67,96 +71,103 @@ public SearchConfig getSearchConfig() {
-			int count = conceptService.getCountOfConceptReferenceTerms(null, conceptSource, context.getIncludeAll());
-			boolean hasMore = count > (context.getStartIndex() + context.getLimit());
-
-			return new AlreadyPaged<ConceptReferenceTerm>(context, terms, hasMore);
+			if (codeOrName != null) {
+				return searchEquals(context, codeOrName, conceptSource);
+			}
+			else {
+				List<ConceptReferenceTerm> terms = conceptService.getConceptReferenceTerms(null, conceptSource,
+						context.getStartIndex(), context.getLimit(), context.getIncludeAll());
+				int count = conceptService.getCountOfConceptReferenceTerms(null, conceptSource, context.getIncludeAll());
+				boolean hasMore = count > (context.getStartIndex() + context.getLimit());
+				return new AlreadyPaged<ConceptReferenceTerm>(context, terms, hasMore);
+			}
+		} else if (searchType.equals(SEARCH_TYPE_STARTS_WITH)) {
+			List<ConceptSource> conceptSources = new ArrayList<ConceptSource>();
+			ConceptSource singleConceptSource = conceptService.getConceptSourceByUuid(source);
+			if (singleConceptSource != null) {
+				conceptSources.add(singleConceptSource);
+			}
+			else {
+				List<ConceptSource> allConceptSources = conceptService.getAllConceptSources();
+				if (allConceptSources.isEmpty()) {
+					return new EmptySearchResult();
+				}
+				else {
+					for (ConceptSource conceptSource : allConceptSources) {
+						if (conceptSource.getName().startsWith(source)) {
+							conceptSources.add(conceptSource);
+						}
+					}
+				}
+			}
+			return searchStartsWith(context, codeOrName, conceptSources);
+		}
+		else {
+			throw new IllegalStateException("\"" + searchType + "\" doesn't fit searchType param requirements. Search type must be " + SEARCH_TYPE_EQUALS + " or " + SEARCH_TYPE_STARTS_WITH + ".");
 		}
 	}
 
-	private PageableResult searchStartsWith(RequestContext context, String code, String name, ConceptSource conceptSource){
+	private PageableResult searchStartsWith(RequestContext context, String codeOrName, List<ConceptSource> conceptSources){
 		/**
 		 * org.openmrs.api.ConceptService#getConceptReferenceTerms(java.lang.String, org.openmrs.ConceptSource, java.lang.Integer, java.lang.Integer, boolean)
 		 * returns all ConceptReferenceTerm objects which code or name match query, so results need to be filtered
 		 */
-		List<ConceptReferenceTerm> resultTerms = new ArrayList<ConceptReferenceTerm>();
-		if(code != null){
-			List<ConceptReferenceTerm> codeQueryTerms = conceptService.getConceptReferenceTerms(code, conceptSource, context.getStartIndex(),
-					context.getLimit(), context.getIncludeAll());
-			for(ConceptReferenceTerm term : codeQueryTerms){
-				if(termMatchesCodeAndName(term, code, name)){
-					resultTerms.add(term);
-				}
+		Set<ConceptReferenceTerm> resultTerms = new LinkedHashSet<ConceptReferenceTerm>();
+		int count = 0;
+
+		if (codeOrName != null && StringUtils.isNotEmpty(codeOrName)) {

```

Comment:
Fixed
PS. Don't merge it yet, since there are some problems with paging when partial source name is querried